toopher/__init__.py

```diff
@@ -238,55 +236,244 @@ def _parse_request_error(self, content):
+        self.update(json_response)
+
     def __nonzero__(self):
-        return self.enabled
+        return self.granted
 
     def __getattr__(self, name):
         if name.startswith('__') or name not in self._raw_data:  # Exclude 'magic' methods to allow for (un)pickling
-            return super(PairingStatus, self).__getattr__(name)
+            return super(AuthenticationRequest, self).__getattr__(name)
         else:
             return self._raw_data[name]
 
+    def authenticate_with_otp(self, api, otp, **kwargs):
+        url = '/authentication_requests/' + self.id + '/otp_auth'
+        params = {'otp' : otp}
+        params.update(kwargs)
+        result = api.advanced.raw.post(url, **params)
+        self.update(result)
+
+    def refresh_from_server(self, api):
+        url = '/authentication_requests/' + self.id
+        result = api.advanced.raw.get(url)
+        self.update(result)
 
-class AuthenticationStatus(object):
-    def __init__(self, json_response):
+    def update(self, json_response):
         try:
             self.id = json_response['id']
             self.pending = json_response['pending']
             self.granted = json_response['granted']
             self.automated = json_response['automated']
             self.reason = json_response['reason']
-
-            terminal = json_response['terminal']
-            self.terminal_id = terminal['id']
-            self.terminal_name = terminal['name']
+            self.terminal.update(json_response['terminal'])
+            self.user.update(json_response['user'])
         except Exception:
             raise ToopherApiError("Could not parse authentication status from response")
 
         self._raw_data = json_response
 
-    def __nonzero__(self):
-        return self.granted
+
+class UserTerminals(object):
+    def __init__(self, raw):
+        self.raw = raw
+
+    def create(self, username, terminal_name, requester_terminal_id, **kwargs):
+        url = '/user_terminals/create'
+        params = {'user_name': username,
+                  'name': terminal_name,
+                  'name_extra': requester_terminal_id}
+        params.update(kwargs)
+        result = self.raw.post(url, **params)
+        return UserTerminal(result)
+
+    def get_by_id(self, terminal_id):
+        url = '/user_terminals/' + terminal_id
+        result = self.raw.get(url)
+        return UserTerminal(result)
+
+
+class UserTerminal(object):
+    def __init__(self, json_response):
+        try:
+            self.user = User(json_response['user'])
+        except Exception:
+            raise ToopherApiError("Could not parse user terminal from response")
+        self.update(json_response)
+
+    def __getattr__(self, name):
+        if name.startswith('__') or name not in self._raw_data:  # Exclude 'magic' methods to allow for (un)pickling
+            return super(UserTerminal, self).__getattr__(name)
+        else:
+            return self._raw_data[name]
```

Comment:
We provide one for `User`, `AuthenticationRequest`, and `Pairing` as well. It looks like it's for if someone asks for an attribute that we return from our API endpoint, but don't explicitly store as an attribute, we can still return it if it's stored in the `_raw_data` (which is a JSON/dict returned by our API request). So if someone wanted to access `auth_request.totp_valid`, even though we don't have it stored as a field in this class explicitly, if we returned it with the API request in `_raw_data`, we could still return the appropriate value.

We saw this being used originally in `Pairing` and `AuthenticationRequest` and so followed that precedent.